File: frontend/src/services/chatService.ts
import { Call } from '@wailsio/runtime'
import * as ChatBinding from '../../bindings/changeme/backend/services/chatservice.js'
import type { Conversation, Message, MessageStatus } from '../types'
import {
    mockConversations,
    mockMessages,
    shouldUseMock,
} from './mockFallback'

/**
 * chatService 统一封装会话/消息数据访问。
 * 所有 UI 组件应通过本 service 获取数据，而非直接调用 Wails bindings。
 *
 * 浏览器独立预览（非 Wails 环境）下自动切换到 mock 数据兜底。
 */

// 浏览器 mock 模式下维护一份会话消息本地副本（支持追加）
const localMockMessages: Record<string, Message[]> = Object.fromEntries(
    Object.entries(mockMessages).map(([k, v]) => [k, [...v]]),
)

// mock 状态推进计时器
function scheduleMockStatus(msg: Message) {
    if (!msg.isSelf) return
    setTimeout(() => applyMockStatus(msg.id, 'sent'), 150)
    setTimeout(() => applyMockStatus(msg.id, 'delivered'), 1200)
}

function applyMockStatus(messageID: string, status: MessageStatus) {
    for (const list of Object.values(localMockMessages)) {
        const idx = list.findIndex((m) => m.id === messageID)
        if (idx >= 0) {
            list[idx] = { ...list[idx], status }
            statusEmitter.emit({
                messageId: messageID,
                conversationId: list[idx].conversationId,
                status,
            })
            return
        }
    }
}

function getSortedMockConversations(): Conversation[] {
    return [...mockConversations].sort((a, b) => {
        if (a.pinned !== b.pinned) return a.pinned ? -1 : 1
        return b.lastTime - a.lastTime
    })
}

const statusEmitter = {
    listeners: new Set<(ev: StatusEvent) => void>(),
    emit(ev: StatusEvent) {
        this.listeners.forEach((fn) => fn(ev))
    },
}

const conversationEmitter = {
    listeners: new Set<() => void>(),
    emit() {
        this.listeners.forEach((fn) => fn())
    },
}

function emitConversationChange() {
    conversationEmitter.emit()
}

export interface StatusEvent {
    messageId: string
    conversationId: string
    status: MessageStatus
}

export const chatService = {
    async getConversations(): Promise<Conversation[]> {
        if (shouldUseMock()) return getSortedMockConversations()
        const result = await ChatBinding.GetConversations()
        return (result ?? []) as unknown as Conversation[]
    },

    async getConversation(id: string): Promise<Conversation | null> {
        if (shouldUseMock())
            return mockConversations.find((c) => c.id === id) ?? null
        const result = await ChatBinding.GetConversation(id)
        return (result ?? null) as unknown as Conversation | null
    },

    async getMessages(conversationId: string, limit = 50): Promise<Message[]> {
        if (shouldUseMock()) {
            const list = localMockMessages[conversationId] ?? []
            return list.slice(-limit)
        }
        const result = await ChatBinding.GetMessages(conversationId, limit)
        return (result ?? []) as unknown as Message[]
    },

    async sendText(
        conversationId: string,
        text: string,
        mentions: string[] = [],
    ): Promise<Message | null> {
        if (shouldUseMock()) {
            const msg: Message = {
                id: `m_${Date.now()}`,
                conversationId,
                senderId: 'u_me',
                senderName: '零零三03',
                senderAvatar: 'https://api.dicebear.com/9.x/adventurer/svg?seed=me',
                type: 'text',
                text: text.trim(),
                timestamp: Date.now(),
                isSelf: true,
                status: 'sending',
                mentions,
            }
            if (!localMockMessages[conversationId]) localMockMessages[conversationId] = []
            localMockMessages[conversationId].push(msg)
            syncMockConversation(conversationId)
            scheduleMockStatus(msg)
            emitConversationChange()
            return msg
        }
        const result = await ChatBinding.SendText(conversationId, text, mentions)
        if (result) emitConversationChange()
        return (result ?? null) as unknown as Message | null
    },

    async sendImage(conversationId: string, mediaUrl: string): Promise<Message | null> {
        if (shouldUseMock()) {
            const msg: Message = mockSelfMsg(conversationId, {
                type: 'image',
                mediaUrl,
            })
            localMockMessages[conversationId] = [
                ...(localMockMessages[conversationId] ?? []),
                msg,
            ]
            syncMockConversation(conversationId)
            scheduleMockStatus(msg)
            emitConversationChange()
            return msg
        }
        const res = await ChatBinding.SendImage(conversationId, mediaUrl)
        if (res) emitConversationChange()
        return (res ?? null) as unknown as Message | null
    },

    async sendVoice(conversationId: string, durationSec: number): Promise<Message | null> {
        if (shouldUseMock()) {
            const msg: Message = mockSelfMsg(conversationId, { type: 'voice', durationSec })
            localMockMessages[conversationId] = [
                ...(localMockMessages[conversationId] ?? []),
                msg,
            ]
            syncMockConversation(conversationId)
            scheduleMockStatus(msg)
            emitConversationChange()
            return msg
        }
        const res = await ChatBinding.SendVoice(conversationId, durationSec)
        if (res) emitConversationChange()
        return (res ?? null) as unknown as Message | null
    },

    async sendFile(
        conversationId: string,
        fileName: string,
        mediaUrl: string,
    ): Promise<Message | null> {
        if (shouldUseMock()) {
            const msg: Message = mockSelfMsg(conversationId, {
                type: 'file',
                fileName,
                mediaUrl,
            })
            localMockMessages[conversationId] = [
                ...(localMockMessages[conversationId] ?? []),
                msg,
            ]
            syncMockConversation(conversationId)
            scheduleMockStatus(msg)
            emitConversationChange()
            return msg
        }
        const res = await ChatBinding.SendFile(conversationId, fileName, mediaUrl)
        if (res) emitConversationChange()
        return (res ?? null) as unknown as Message | null
    },

    async sendLocation(conversationId: string, title: string): Promise<Message | null> {
        if (shouldUseMock()) {
            const msg: Message = mockSelfMsg(conversationId, { type: 'location', text: title })
            localMockMessages[conversationId] = [
                ...(localMockMessages[conversationId] ?? []),
                msg,
            ]
            syncMockConversation(conversationId)
            scheduleMockStatus(msg)
            emitConversationChange()
            return msg
        }
        const res = await ChatBinding.SendLocation(conversationId, title)
        if (res) emitConversationChange()
        return (res ?? null) as unknown as Message | null
    },

    async markRead(conversationId: string): Promise<void> {
        if (shouldUseMock()) {
            const list = localMockMessages[conversationId] ?? []
            list.forEach((m, i) => {
                if (m.isSelf && m.status !== 'read' && m.status !== 'failed') {
                    list[i] = { ...m, status: 'read' }
                }
            })
            // mock 也要清 unread
            const idx = mockConversations.findIndex((c) => c.id === conversationId)
            if (idx >= 0) mockConversations[idx].unreadCount = 0
            emitConversationChange()
            return
        }
        await ChatBinding.MarkRead(conversationId)
        emitConversationChange()
    },

    async setPinned(conversationId: string, pinned: boolean): Promise<void> {
        if (shouldUseMock()) {
            const idx = mockConversations.findIndex((c) => c.id === conversationId)
            if (idx >= 0) mockConversations[idx].pinned = pinned
            emitConversationChange()
            return
        }
        await Call.ByName('changeme/backend/services.ChatService.SetPinned', conversationId, pinned)
        emitConversationChange()
    },

    async setMute(conversationId: string, mute: boolean): Promise<void> {
        if (shouldUseMock()) {
            const idx = mockConversations.findIndex((c) => c.id === conversationId)
            if (idx >= 0) mockConversations[idx].muteNotice = mute
            emitConversationChange()
            return
        }
        await Call.ByName('changeme/backend/services.ChatService.SetMute', conversationId, mute)
        emitConversationChange()
    },

    async deleteConversation(conversationId: string): Promise<void> {
        if (shouldUseMock()) {
            const idx = mockConversations.findIndex((c) => c.id === conversationId)
            if (idx >= 0) mockConversations.splice(idx, 1)
            delete localMockMessages[conversationId]
            emitConversationChange()
            return
        }
        await Call.ByName('changeme/backend/services.ChatService.DeleteConversation', conversationId)
        emitConversationChange()
    },

    async recallMessage(messageId: string): Promise<void> {
        if (shouldUseMock()) {
            for (const list of Object.values(localMockMessages)) {
                const idx = list.findIndex((m) => m.id === messageId)
                if (idx >= 0) {
                    const conversationId = list[idx].conversationId
                    const name = list[idx].senderName || '成员'
                    list[idx] = {
                        ...list[idx],
                        type: 'tip',
                        text: `${name} 撤回了一条消息`,
                        mediaUrl: undefined,
                        fileName: undefined,
                        status: undefined,
                        isSelf: false,
                    }
                    syncMockConversation(conversationId)
                    emitConversationChange()
                    return
                }
            }
            return
        }
        await ChatBinding.RecallMessage(messageId)
        emitConversationChange()
    },

    async deleteMessage(messageId: string): Promise<void> {
        if (shouldUseMock()) {
            for (const [cid, list] of Object.entries(localMockMessages)) {
                const idx = list.findIndex((m) => m.id === messageId)
                if (idx >= 0) {
                    list.splice(idx, 1)
                    // 同步 mock 会话列表最后一条
                    const convIdx = mockConversations.findIndex((c) => c.id === cid)
                    if (convIdx >= 0 && list.length > 0) {
                        const last = list[list.length - 1]
                        mockConversations[convIdx].lastMessage = last.text ?? ''
                        mockConversations[convIdx].lastTime = last.timestamp
                    } else if (convIdx >= 0) {
                        mockConversations[convIdx].lastMessage = ''
                    }
                    emitConversationChange()
                    return
                }
            }
            return
        }
        await ChatBinding.DeleteMessage(messageId)
        emitConversationChange()
    },

    async setDraft(conversationId: string, draft: string): Promise<void> {
        if (shouldUseMock()) {
            const idx = mockConversations.findIndex((c) => c.id === conversationId)
            if (idx >= 0) mockConversations[idx].draft = draft
            emitConversationChange()
            return
        }
        await ChatBinding.SetDraft(conversationId, draft)
        emitConversationChange()
    },

    async createGroup(title: string, memberIds: string[]): Promise<Conversation | null> {
        if (shouldUseMock()) {
            const id = `c_group_${Date.now()}`
            const conv: Conversation = {
                id,
                type: 'group',
                title: title || `群聊(${memberIds.length + 1})`,
                avatarUrl: '',
                lastMessage: '群聊已创建。欢迎开始聊天～',
                lastTime: Date.now(),
                unreadCount: 0,
                pinned: false,
                muteNotice: false,
                memberIds: ['u_me', ...memberIds],
                ownerId: 'u_me',
                memberAvatars: [],
            }
            mockConversations.unshift(conv)
            localMockMessages[id] = [
                {
                    id: `m_${Date.now()}_tip`,
                    conversationId: id,
                    senderId: '',
                    senderName: '',
                    senderAvatar: '',
                    type: 'tip',
                    text: '群聊已创建。欢迎开始聊天～',
                    timestamp: Date.now(),
                    isSelf: false,
                },
            ]
            emitConversationChange()
            return conv
        }
        const res = await ChatBinding.CreateGroup(title, memberIds)
        if (res) emitConversationChange()
        return (res ?? null) as unknown as Conversation | null
    },

    notifyConversationChange(): void {
        emitConversationChange()
    },

    onConversationChange(handler: () => void): () => void {
        conversationEmitter.listeners.add(handler)
        return () => conversationEmitter.listeners.delete(handler)
    },

    onStatus(handler: (ev: StatusEvent) => void): () => void {
        if (shouldUseMock()) {
            statusEmitter.listeners.add(handler)
            return () => statusEmitter.listeners.delete(handler)
        }
        // Wails 自动生成类型里的 MessageStatus 是 enum，与我们 union 类型不完全兼容，
        // 所以在订阅边界上用 any 手动转换，前端内部继续使用清晰的 union。
        let unsubscribe = () => { }
        import('@wailsio/runtime').then((m) => {
            unsubscribe = m.Events.On('chat:status', (ev: { data?: unknown }) => {
                handler(ev.data as StatusEvent)
            })
        })
        return () => unsubscribe()
    },
}

function mockSelfMsg(conversationId: string, extra: Partial<Message>): Message {
    return {
        id: `m_${Date.now()}`,
        conversationId,
        senderId: 'u_me',
        senderName: '零零三03',
        senderAvatar: 'https://api.dicebear.com/9.x/adventurer/svg?seed=me',
        type: 'text',
        timestamp: Date.now(),
        isSelf: true,
        status: 'sending',
        ...extra,
    }
}

function syncMockConversation(conversationId: string) {
    const idx = mockConversations.findIndex((c) => c.id === conversationId)
    if (idx < 0) return
    const list = localMockMessages[conversationId] ?? []
    const last = list[list.length - 1]
    if (!last) return
    mockConversations[idx].lastMessage = getMessagePreview(last)
    mockConversations[idx].lastTime = last.timestamp
}

function getMessagePreview(message: Message): string {
    switch (message.type) {
        case 'image':
            return '[图片]'
        case 'video':
            return '[视频]'
        case 'voice':
            return '[语音]'
        case 'call':
            return '[音视频通话]'
        case 'file':
            return message.fileName ? `[文件] ${message.fileName}` : '[文件]'
        case 'location':
            return '[位置]'
        case 'redpacket':
            return '[红包]'
        case 'tip':
            return message.text ?? ''
        default:
            return message.text ?? ''
    }
}

File: backend/services/chat_service.go
package services

import (
	"fmt"
	"strings"
	"time"

	"changeme/backend/domain"
	"changeme/backend/repository"

	"github.com/wailsapp/wails/v3/pkg/application"
)

// MessageStatusEvent 消息状态变更事件，前端订阅 "chat:status" 实时更新"已发送/已读"
type MessageStatusEvent struct {
	MessageID      string               `json:"messageId"`
	ConversationID string               `json:"conversationId"`
	Status         domain.MessageStatus `json:"status"`
}

// ChatService 提供会话/消息相关的前端调用接口
type ChatService struct {
	repo repository.ConversationRepository
}

func NewChatService(r repository.ConversationRepository) *ChatService {
	return &ChatService{repo: r}
}

// GetConversations 返回会话列表（倒序）
func (s *ChatService) GetConversations() ([]domain.Conversation, error) {
	return s.repo.ListConversations()
}

// GetConversation 根据 ID 获取单个会话
func (s *ChatService) GetConversation(id string) (*domain.Conversation, error) {
	return s.repo.GetConversation(id)
}

// GetMessages 获取某会话的最新 limit 条消息（按时间正序）
func (s *ChatService) GetMessages(conversationID string, limit int) ([]domain.Message, error) {
	return s.repo.ListMessages(conversationID, limit)
}

// SendText 发送文本消息。初始 status=sending；两个计时器模拟 sent→delivered→read。
// mentions: 被 @ 的 userID/botID 列表。
func (s *ChatService) SendText(
	conversationID string,
	text string,
	mentions []string,
) (*domain.Message, error) {
	trimmed := strings.TrimSpace(text)
	if trimmed == "" {
		return nil, fmt.Errorf("text cannot be empty")
	}
	msg := s.buildSelfMessage(conversationID, domain.MsgText)
	msg.Text = trimmed
	msg.Mentions = mentions
	if err := s.repo.AppendMessage(msg); err != nil {
		return nil, err
	}
	// 异步推进状态：sending -> sent -> delivered（模拟对方收到）
	go s.advanceStatus(msg.ID, conversationID)
	return &msg, nil
}

// SendImage 发送图片消息
func (s *ChatService) SendImage(conversationID, mediaURL string) (*domain.Message, error) {
	msg := s.buildSelfMessage(conversationID, domain.MsgImage)
	msg.MediaURL = mediaURL
	if err := s.repo.AppendMessage(msg); err != nil {
		return nil, err
	}
	go s.advanceStatus(msg.ID, conversationID)
	return &msg, nil
}

// SendVoice 语音
func (s *ChatService) SendVoice(conversationID string, durationSec int) (*domain.Message, error) {
	msg := s.buildSelfMessage(conversationID, domain.MsgVoice)
	msg.DurationSec = durationSec
	if err := s.repo.AppendMessage(msg); err != nil {
		return nil, err
	}
	go s.advanceStatus(msg.ID, conversationID)
	return &msg, nil
}

// SendFile 文件
func (s *ChatService) SendFile(conversationID, fileName, mediaURL string) (*domain.Message, error) {
	msg := s.buildSelfMessage(conversationID, domain.MsgFile)
	msg.FileName = fileName
	msg.MediaURL = mediaURL
	if err := s.repo.AppendMessage(msg); err != nil {
		return nil, err
	}
	go s.advanceStatus(msg.ID, conversationID)
	return &msg, nil
}

// SendLocation 位置
func (s *ChatService) SendLocation(conversationID, title string) (*domain.Message, error) {
	msg := s.buildSelfMessage(conversationID, domain.MsgLoc)
	msg.Text = title
	if err := s.repo.AppendMessage(msg); err != nil {
		return nil, err
	}
	go s.advanceStatus(msg.ID, conversationID)
	return &msg, nil
}

// MarkRead 进入会话详情时清零未读，并把自己过往消息标记为 read（仅影响 UI "已读"标志）。
func (s *ChatService) MarkRead(conversationID string) error {
	return s.repo.MarkConversationRead(conversationID)
}

func (s *ChatService) SetPinned(conversationID string, pinned bool) error {
	return s.repo.SetConversationPinned(conversationID, pinned)
}

func (s *ChatService) SetMute(conversationID string, mute bool) error {
	return s.repo.SetConversationMute(conversationID, mute)
}

func (s *ChatService) DeleteConversation(conversationID string) error {
	return s.repo.DeleteConversation(conversationID)
}

// CreateGroup 创建群聊。memberIDs 可混合 u_* 与 bot_* ID。
func (s *ChatService) CreateGroup(title string, memberIDs []string) (*domain.Conversation, error) {
	return s.repo.CreateGroup(title, "u_me", memberIDs)
}

// SetDraft 保存输入框草稿。
func (s *ChatService) SetDraft(conversationID, draft string) error {
	return s.repo.SetDraft(conversationID, draft)
}

// RecallMessage 撤回自己发送的消息，将其转成 tip "XX 撤回了一条消息"
func (s *ChatService) RecallMessage(messageID string) error {
	return s.repo.RecallMessage(messageID)
}

// DeleteMessage 删除消息（本地数据层删除）
func (s *ChatService) DeleteMessage(messageID string) error {
	return s.repo.DeleteMessage(messageID)
}

// buildSelfMessage 通用的"我的"消息骨架。
func (s *ChatService) buildSelfMessage(conversationID string, t domain.MessageType) domain.Message {
	now := time.Now()
	return domain.Message{
		ID:             fmt.Sprintf("m_%d", now.UnixNano()),
		ConversationID: conversationID,
		SenderID:       "u_me",
		SenderName:     "零零三03",
		SenderAvatar:   "https://api.dicebear.com/9.x/adventurer/svg?seed=me",
		Type:           t,
		Timestamp:      now.UnixMilli(),
		IsSelf:         true,
		Status:         domain.StatusSending,
	}
}

// advanceStatus 模拟消息状态流转：100ms sent -> 1.2s delivered。
// "read" 由机器人回复完成 或 用户切到该聊天 时触发。
func (s *ChatService) advanceStatus(messageID, conversationID string) {
	time.Sleep(100 * time.Millisecond)
	_ = s.repo.UpdateMessageStatus(messageID, domain.StatusSent)
	s.emitStatus(messageID, conversationID, domain.StatusSent)
	time.Sleep(1 * time.Second)
	_ = s.repo.UpdateMessageStatus(messageID, domain.StatusDelivered)
	s.emitStatus(messageID, conversationID, domain.StatusDelivered)
}

func (s *ChatService) emitStatus(messageID, conversationID string, status domain.MessageStatus) {
	app := application.Get()
	if app == nil {
		return
	}
	app.Event.Emit("chat:status", MessageStatusEvent{
		MessageID:      messageID,
		ConversationID: conversationID,
		Status:         status,
	})
}
